### PR TITLE
Re-enable, add and fix tests for Templates / Refactor Template.js Class

### PR DESCRIPTION
**Pull request by james-wallis**
### Summary 
Fix for https://github.com/eclipse/codewind/issues/1207 (Enable template tests)
#### Template Tests
* Re-enable existing API and Unit tests for Templates
* Fix the tests which have broken due to changes while they were deactivated
* Add tests for functions which previously did not have tests
  * Coverage for Template.js is now above 90%
* Removed Appsody templates from the unit and API tests (they use the Codewind repository as the testing repository as we are in total control of when it is updated whereas Appsody can be updated at any time).
* Refactored the tests according to the changes and layout of Template.js. 
  * Split the tests into two sections (Class and Local functions) to make the tests clearer.
  * Each test has been ordered matching the layout of the functions in Template.js to make the running of the tests streamlined.

#### Refactoring Template.js
* Major changes
  * Arranged the functions inside the Template class object into three sections - templates, repositories and providers in order group the functions and improve code readability.
  * Moved functions out of the class into local functions as they were only called in the Templates file to improve code readability. Also improves the code layout as it is easier to see what a function updates rather than it updating `this` inside it.
  * Removed the `updatePromise` from the Template routes and moved its functionality into Template.js as below:
    * Added `projectRepositoriesNeedsRefresh` which will be set to true each time a provider (such as Appsody) is added. This will update the `repositoryList` when a `getRepositories` or `getTemplates` are called and set `projectRepositoriesNeedsRefresh` to false.

* Minor changes
  * Removed duplicated code in places, for example restructured `initializeRepositoryList` and created the `getJSONFromURL` function.
  * Changed `needsRefresh` to `projectTemplatesNeedsRefresh` to be more explicit in its function. Also made it so it is used to refresh the templates when a repository is added or deleted it will be set to true and on `getTemplates` the templates list will be updated and `projectTemplatesNeedsRefresh` set to false. Enabling this flag should improve performance.
  * Changed `getTemplates` to take in two parameters rather than an object.
  * Made `getRepositories` an async function to `async/await` the `updateRepoListWithReposFromProviders` function, this led to more code being `async/await`. While doing this I added `await`s  to functions which should be awaited but were not.
  * Removed module.exports for testing and used `rewire` instead, its cleaner.

#### Other
* Change projectTypes API to call getEnabledTemplates with a `projectType` of false.

### CODE REVIEW

File: test/src/unit/template.test.js [hunk 1]
```diff
@@ -53,705 +55,519 @@ const mockRepos = {
...
             });
-        });
-        describe('only Appsody templates', function() {
-            const sampleTemplateList = [sampleAppsodyTemplate];
-            let templateController;
-            before(() => {
-                templateController = new Templates('');
-                templateController.projectTemplates = sampleTemplateList;
-                templateController.needsRefresh = false;
+            it('initialises a Templates Class and creates a repository file', async function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: consistency with initialize vs initialise

File: test/src/unit/template.test.js [hunk 2]
```diff
@@ -53,705 +55,519 @@ const mockRepos = {
...
-            let templateController;
-            before(async() => {
-                fs.ensureDirSync(testWorkspaceConfigDir);
-                templateController = new Templates(testWorkspaceDir);
-                await templateController.writeRepositoryList();
-                templateController.addProvider('valid repo', {
-                    getRepositories() {
-                        return [sampleRepos.appsody];
-                    },
+            describe('Verifys the Codewind default templates are there and all others contain the required keys', function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: Verifies?

File: test/src/unit/template.test.js [hunk 3]
```diff
@@ -53,705 +55,519 @@ const mockRepos = {
...
             });
-            after(() => {
-                fs.removeSync(testWorkspaceDir);
-            });
-            it('returns the default templates and more', async function() {
-                this.timeout = 10000;
-                const output = await templateController.getAllTemplates();
-                output.should.include.deep.members(defaultCodewindTemplates);
-                (output.length).should.be.above(defaultCodewindTemplates.length);
+            describe('and add an extra bad template repo', function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: 'extra bad' sounds like it's a really bad template, could be `additional invalid template`?

File: test/src/unit/template.test.js [hunk 4]
```diff
@@ -53,705 +55,519 @@ const mockRepos = {
...
-                    description: 'not a GitHub repo',
-                }],
-            },
-        };
-        for (const [testName, test] of Object.entries(tests)) {
-            describe(testName, function() {
-                it(`returns the expected repos`, async function() {
-                    const output = await Templates.getReposFromProviders(test.input);
-                    output.should.deep.equal(test.output);
+                it('returns the default templates and more', async function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: more is a bit vague ... could say default templates, and extra non-default templates

File: test/src/unit/template.test.js [hunk 5]
```diff
@@ -53,705 +55,519 @@ const mockRepos = {
...
-        });
-        describe('(existing url)', function() {
-            it('enables the correct repo', function() {
-                templateController.enableRepository(mockRepos.disabled.url);
-                const expectedRepoDetails = {
-                    ...mockRepos.disabled,
-                    enabled: true,
-                };
-                templateController.getRepositories().should.deep.include(expectedRepoDetails);
+            it('Returns true as the repository exists', async function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: inconsistent casing

File: test/src/unit/template.test.js [hunk 6]
```diff
@@ -760,198 +576,647 @@ describe('Templates.js', function() {
...
+                const func = () => getTemplatesFromRepo({ url: 'invalidURL' });
+                return func().should.be.rejectedWith('Invalid URL');
+            });
+            it('throws a useful error when a valid URL is given but it does not point to JSON', function() {
+                const func = () => getTemplatesFromRepo({ url: 'https://www.google.com/' });
+                return func().should.be.rejectedWith(`URL 'https://www.google.com/' should return JSON`);
+            });
+            it('throws a useful error when an invalid file path is given', function() {
+                const func = () => getTemplatesFromRepo({ url: 'file://someting.json' });
+                return func().should.be.rejectedWith(`repo file 'file://someting.json/' should return json`);
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: Sometimg

File: test/src/unit/template.test.js (hunk 2 above)
Comment (jcockbain) [reply test/src/unit/template.test.js]: Also lower case for consistency

File: test/src/unit/template.test.js (hunk 3 above)
Comment (jcockbain) [reply test/src/unit/template.test.js]: I'd also get rid of the and from this, and the below describe

File: test/src/unit/template.test.js [hunk 7]
```diff
@@ -760,198 +576,647 @@ describe('Templates.js', function() {
...
+                output.length.should.equal(0);
+            });
+        });
+        describe('getTemplateStyles(templates)', function() {
+            const getTemplateStyles = Templates.__get__('getTemplateStyles');
+            it('returns Codewind by default', function() {
+                const output = getTemplateStyles([mockRepos.enabled]);
+                output.should.deep.equal(['Codewind']);
+            });
+            it('returns Codewind as the templates are Codewind ones', function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: this message is a bit confusing

File: test/src/unit/template.test.js [hunk 8]
```diff
@@ -760,198 +576,647 @@ describe('Templates.js', function() {
...
             });
-        }
-    });
-    describe('filterTemplatesByStyle(templates, projectStyle)', function() {
-        const templates = [sampleCodewindTemplate, sampleAppsodyTemplate];
-        describe(`projectStyle='Codewind'`, function() {
-            it('returns only Codewind templates', function() {
-                const output = Templates.filterTemplatesByStyle(templates, 'Codewind');
-                output.should.deep.equal([sampleCodewindTemplate]);
+            it('Returns false as the URL does not point to JSON', function() {
```
Comment (jcockbain) [inline test/src/unit/template.test.js]: I'd do lower case for consistency

File: test/src/unit/template.test.js [hunk 9]
```diff
@@ -53,905 +55,1186 @@ const mockRepos = {
...
-                describe(testName, function() { // eslint-disable-line no-loop-func
-                    it(`returns the expected operation info and correctly updates the repository file`, async function() {
-                        const output = await templateController.batchUpdate(test.input);
-                        output.should.deep.equal(test.output);
+            for (const [title, test] of Object.entries(tests)) {
+                const { input, output } = test;
+                // eslint-disable-next-line no-loop-func
+                it(`returns 200 and ${title}`, async() => {
+                    const templateController = new Templates('');
+                    const testRepo = (output) ? mockRepos.disabled : mockRepos.enabled;
```
Comment (rwalle61) [inline test/src/unit/template.test.js]: split

File: src/pfe/portal/modules/Templates.js [hunk 10]
```diff
@@ -52,153 +52,110 @@ module.exports = class Templates {
...
+    let templates = [...this.projectTemplates];
+    if (this.projectTemplatesNeedsRefresh) {
+      const repositories = (String(showEnabledOnly) === 'true')
+        ? await this.getEnabledRepositories()
+        : this.repositoryList;
+      templates = await getTemplatesFromRepos(repositories);
+      this.projectTemplates = templates;
+      this.projectTemplatesNeedsRefresh = false;
     }
+    if (projectStyle) return filterTemplatesByStyle(templates, projectStyle);
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: I've added this function as 
```
async getTemplatesByStyle(projectStyle, showEnabledOnly = false) {
    const templates = await this.getTemplates(showEnabledOnly);
    return filterTemplatesByStyle(templates, projectStyle);
  }
```
and removed the projectStyle parameter from getTemplates.

File: src/pfe/portal/modules/Templates.js [hunk 11]
```diff
@@ -52,153 +52,110 @@ module.exports = class Templates {
...
-      if (await cwUtils.fileExists(this.repositoryFile)) {
-        this.repositoryList = await fs.readJson(this.repositoryFile); // eslint-disable-line require-atomic-updates
-        await this.updateRepoListWithReposFromProviders();
-        this.repositoryList = await fetchAllRepositoryDetails(this.repositoryList); // eslint-disable-line require-atomic-updates
-        this.needsRefresh = true;
-      } else {
-        await this.updateRepoListWithReposFromProviders();
-        this.repositoryList = await fetchAllRepositoryDetails(this.repositoryList); // eslint-disable-line require-atomic-updates
-        await this.writeRepositoryList();
+      let repositories = [...this.repositoryList];
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: Not adding as this function shouldn't change that at the moment. If functionality is added in the future as you say, it should be added then.

File: src/pfe/portal/modules/Templates.js [hunk 12]
```diff
@@ -52,153 +52,110 @@ module.exports = class Templates {
...
-          repo.protected = true;
-          const repoWithTemplateStyles = await fetchRepositoryDetails(repo);
-          return repoWithTemplateStyles;
-        })
-      );
-      this.repositoryList = this.repositoryList.concat(reposWithCodewindSettings);
-      await this.writeRepositoryList();
+    let templates = [...this.projectTemplates];
+    if (this.projectTemplatesNeedsRefresh) {
+      const repositories = (String(showEnabledOnly) === 'true')
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: Simplified.

File: src/pfe/portal/modules/Templates.js [hunk 13]
```diff
@@ -52,153 +52,110 @@ module.exports = class Templates {
...
-  }
+  // TEMPLATES
 
-  async getAllTemplates() {
-    if (!this.needsRefresh) {
-      return this.projectTemplates;
+  async getTemplates(showEnabledOnly, projectStyle) {
+    if (this.projectRepositoriesNeedsRefresh) {
+      // eslint-disable-next-line require-atomic-updates
+      this.repositoryList = await updateRepoListWithReposFromProviders(this.providers, this.repositoryList, this.repositoryFile);
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: Changed based on your comments.

File: src/pfe/portal/modules/Templates.js [hunk 14]
```diff
@@ -52,153 +52,110 @@ module.exports = class Templates {
...
-    this.projectTemplates = newProjectTemplates;
-    return this.projectTemplates;
-  }
-
-  // Save the default list to disk so the user can potentially edit it (WHEN CODEWIND IS NOT RUNNING)
-  async writeRepositoryList() {
-    await fs.writeJson(this.repositoryFile, this.repositoryList, { spaces: '  ' });
-    log.info(`Repository list updated.`);
+  getEnabledTemplates(projectStyle) {
+    return this.getTemplates(true, projectStyle);
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: removed this function

File: src/pfe/portal/modules/Templates.js [hunk 15]
```diff
@@ -52,153 +52,110 @@ module.exports = class Templates {
...
+    } catch (error) {
+      return false;
+    }
   }
 
-  async batchUpdate(requestedOperations) {
-    const operationResults = requestedOperations.map(operation => this.performOperation(operation));
-    await this.writeRepositoryList();
+  async batchUpdate(requestedOperations) { 
+    const promiseArray = requestedOperations.map(operation => this.performOperationOnRepository(operation));
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: changed to promiseList

File: src/pfe/portal/modules/Templates.js [hunk 16]
```diff
@@ -310,55 +264,80 @@ module.exports = class Templates {
...
+
+async function updateRepoListWithReposFromProviders(providers, repositoryList, repositoryFile) {
+  const providedRepos = await getReposFromProviders(Object.values(providers));
+  
+  const extraRepos = providedRepos.filter(repo =>
+    !repositoryList.find(repo2 => repo2.url === repo.url)
+  );
+  if (extraRepos.length > 0) {
+    
+    const reposWithCodewindSettings = await Promise.all(
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: moved into `addCodewindSettingsToRepository` function

File: src/pfe/portal/routes/projectTypes.route.js [hunk 17]
```diff
@@ -83,7 +83,7 @@ router.get('/api/v1/project-types', async (req, res) => {
   const projectTypes = [];
   const seenProjectTypes = {};
   try {
-    const templates = await user.templates.getEnabledTemplates();
+    const templates = await user.templates.getEnabledTemplates(false);
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/routes/projectTypes.route.js]: changed this

File: src/pfe/portal/modules/Templates.js [hunk 18]
```diff
@@ -410,61 +387,48 @@ async function readRepoTemplatesJSON(repository) {
...
   if (res.statusCode !== 200) {
-    // Return as templates.json may not exist.
-    return;
+    // Return repository untouched as templates.json may not exist.
+    return repository;
   }
 
   try {
     const templateDetails = JSON.parse(res.body);
+    const updatedRepository = repository;
```
[the remark this answers is not in the document]
Comment (james-wallis) [reply src/pfe/portal/modules/Templates.js]: Changed

**Comment by james-wallis** [comment]
In addition to the PR comments above, I've also changed the `delete repository` API function to return a 404 if the project does not exist. Currently we just return `200`.

**Pull request merged**